text/0000-derive-deref.md

```diff
@@ -0,0 +1,47 @@
+- Feature Name: derive-deref
+- Start Date: 2016-08-01
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow deriving `Deref` and `DerefMut`.
+
+# Motivation
+[motivation]: #motivation
+
+Newtypes are a common pattern in Rust, and dereferencing is a convenient way to
+work with these. Implementing `Deref` and `DerefMut` is quite mildly annoying,
+compared to the triviality.
+
+# Detailed design
+[design]: #detailed-design
+
+Add a `derive_Deref` and `derive_DerefMut` attribute (that is, `derive(Deref)`
+and `derive(DerefMut)`), which derives the respective implementation for
+single-field `struct`s to deref to their only field.
+
+```rust
+#[derive(Deref, DerefMut)]
+struct MyType<T> {
+    inner: T,
+}
+
+// Now MyType<T> implements Deref<Target = T> as well as DerefMut.
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+None.
+
+# Alternatives
+[alternatives]: #alternatives
+
+None.
```

Comment:
Submitting an RFC with no drawbacks and no alternatives makes it very likely to be closed without the attention it might otherwise deserve.

In this case newtype deriving seems to be the main alternative. Perhaps also support for custom derives would allow this to be in a crate, rather than part of the language.

A major disadvantage I see is that this encourages using `Deref` with non-pointer types. It might be that newtypes (like slices/collections) are a reasonable use of `Deref` (c.f., using `Deref` for polymorphism which I regard as an anti-pattern, and which this is dangerously close to). It would be good to see that argument made in this RFC.